### PR TITLE
uuid for obstacle message

### PR DESCRIPTION
**Pull request by simutisernestas**
don't really like the solution of the tracking of those Rviz markers ids like that, but can't think of a better way to do it for now

### CODE REVIEW

File: kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py [hunk 1]
```diff
@@ -165,13 +166,14 @@ def callback(self, msg):
         if self.tracker_marker_pub.get_subscription_count() > 0:
             marker_array = MarkerArray()
             marker_list = []
+            self.current_rviz_marker_id += 1 # TODO collision
             # add current active obstacles
             for obs in filtered_obstacle_list:
-                (r, g, b) = colorsys.hsv_to_rgb(obs.msg.id * 30. % 360 / 360., 1., 1.) # encode id with rgb color
+                (r, g, b) = colorsys.hsv_to_rgb(i * 30. % 360 / 360., 1., 1.) # encode id with rgb color
```
Comment (SteveMacenski) [inline kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py]: Where is "`i`" defined?

File: kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py [hunk 2]
```diff
@@ -185,9 +187,10 @@ def callback(self, msg):
                 marker.scale = obs.msg.size
                 marker_list.append(marker)
                 # make an arrow
+                arrow_id = 2**32 - self.current_rviz_marker_id # descending from 32 bit id
                 arrow = Marker()
                 arrow.header = msg.header
-                arrow.id = 255 - obs.msg.id
+                arrow.id = arrow_id
```
Comment (SteveMacenski) [inline kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py]: This is for visualization - my memory on `Marker` is a little wavering, but don't you have the "namespace" you can use to make a marker into `ns/id`? If so, you can make the `id` to `0` always and have the `namespace` by the `uuid` of the message. Then there's no 32 bit shifting or marker maps or any of that stuff. Test to make sure though.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py (hunk 2 above)
Comment (simutisernestas) [reply kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py]: Yes, you are right. ```Marker``` has a namespace field. Thanks

File: kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py [hunk 3]
```diff
@@ -167,11 +165,12 @@ def callback(self, msg):
             marker_list = []
             # add current active obstacles
             for obs in filtered_obstacle_list:
-                (r, g, b) = colorsys.hsv_to_rgb(obs.msg.id * 30. % 360 / 360., 1., 1.) # encode id with rgb color
+                (r, g, b) = colorsys.hsv_to_rgb((obs.msg.id.int >> 64) * 30. % 360 / 360., 1., 1.) # encode id with rgb color
```
Comment (SteveMacenski) [inline kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py]: When I run this in ipython a few times, its always gives me out the same 3 values: 0.25, 0.916, and 0.58.

```
(uuid.uuid1().int >> 64) * 30 % 360 / 360
```

I don't think this works.

**Comment by simutisernestas** [comment]
i haven't tested it yet on real data, i'll ping you after I do that

**Comment by SteveMacenski** [comment]
Those changes make 3 options to ~11 options. Definitely an improvement, but not really randomly distributed. 

I did some playing `uuid.uuid1().int % 100` will give you 25 evenly distributed numbers between 0 and 100 (so divide by 100). From 1000 iterations:
```
{81: 40,
 17: 40,
 53: 40,
 89: 40,
 25: 40,
 61: 40,
 97: 40,
 33: 40,
 69: 40,
 5: 40,
 41: 40,
 77: 40,
 13: 40,
 49: 40,
 85: 40,
 21: 40,
 57: 40,
 93: 40,
 29: 40,
 65: 40,
 1: 40,
 37: 40,
 73: 40,
 9: 40,
 45: 40}
```

Each of those values appears exactly 40 times. Still not random, but closer.

**Comment by simutisernestas** [comment]
`uuid.uuid4().int % 100` gives all 100 distributed like this:
```
31: 12
57: 13
18: 11
1: 9
98: 7
96: 5
6: 18
44: 13
81: 9
3: 9
10: 13
11: 11
93: 12
68: 13
55: 8
91: 13
67: 14
26: 15
41: 11
8: 12
94: 8
61: 16
25: 7
99: 14
43: 9
37: 10
23: 17
73: 15
63: 9
85: 9
20: 7
75: 11
35: 15
33: 8
84: 13
58: 11
74: 7
32: 4
72: 5
70: 9
59: 10
80: 14
64: 7
83: 11
34: 10
60: 13
21: 8
90: 7
28: 11
65: 7
54: 8
50: 12
92: 10
4: 11
56: 11
53: 6
86: 11
89: 8
16: 18
78: 7
87: 12
48: 8
97: 10
95: 6
39: 9
27: 7
69: 11
66: 8
51: 11
79: 4
71: 9
49: 7
47: 8
9: 12
46: 8
42: 10
29: 12
62: 12
12: 12
45: 8
88: 15
13: 5
30: 15
15: 12
77: 7
22: 15
14: 10
38: 10
76: 11
2: 5
52: 5
82: 5
7: 9
40: 7
19: 12
36: 9
0: 8
24: 7
17: 11
5: 6
```
Looks much better. I've thought that `uuid1` will be more random because of the time component. Should I use this?

**Comment by SteveMacenski** [comment]
Use uuid4, googling around makes me think this is a good tradeoff.

**Review by SteveMacenski — APPROVED** [review]
@tony23545 please review

**Comment by simutisernestas** [comment]
I've found some major bugs while testing, so there will be quite some changes

**Comment by simutisernestas** [comment]
Now it's tested and working properly. The first implementation was too naive. Assigning python UUID object directly to ROS msg was not a good idea :). Script below should explain those changes better: 
```
import uuid 
from unique_identifier_msgs.msg import UUID

if __name__ == '__main__':
    raw_uuid = uuid.uuid4()
    print(raw_uuid)

    uuid_msg = UUID()
    print(raw_uuid.bytes)

    uuid_msg.uuid = list(raw_uuid.bytes)
    print(uuid_msg.uuid)
    print(bytes(uuid_msg.uuid))

    a = uuid.UUID(bytes=bytes(uuid_msg.uuid))
    print(a)
```
Output:
 ```
b9db0cd7-68a4-4079-9228-2b8991f57def
b'\xb9\xdb\x0c\xd7h\xa4@y\x92(+\x89\x91\xf5}\xef'
[185 219  12 215 104 164  64 121 146  40  43 137 145 245 125 239]
b'\xb9\xdb\x0c\xd7h\xa4@y\x92(+\x89\x91\xf5}\xef'
b9db0cd7-68a4-4079-9228-2b8991f57def

**Review by SteveMacenski — APPROVED** [review]
@tony23545 please review

**Comment by tony23545** [comment]
It looks good to me.

**Comment by SteveMacenski** [comment]
@simutisernestas any other changes before I can merge?

**Comment by simutisernestas** [comment]
GTG already

**Pull request merged**